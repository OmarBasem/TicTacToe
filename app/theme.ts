export const theme = {
    xColor: 'rgb(239, 25, 27)',
    xColorDim: 'rgba(239, 25, 27, 0.5)',
    oColor: 'rgb(31, 237, 238)',
    oColorDim: 'rgba(31, 237, 238, 0.5)'
}