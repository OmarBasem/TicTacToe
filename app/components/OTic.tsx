import React from 'react';
import {Text, StyleSheet, Platform} from "react-native";

import {W} from '../utils';
import {theme} from "../theme";

const OTic = () => {
    return (
        <Text style={s.x}>O</Text>
    )
}

const s = StyleSheet.create({
    x: {
        color: theme.oColor,
        fontSize: W(25),
        bottom: Platform.OS === 'ios' ? 4 : 16
    }
})

export default OTic;