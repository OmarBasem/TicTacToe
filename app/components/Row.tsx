import React from 'react'
import {View, StyleSheet, Text} from 'react-native';

import Box from './Box'

const Row = ({rIndex}: {rIndex: number}) => {
    const col = [0,1,2]
    return (
        <View style={s.row}>
            {col.map(cIndex => <Box key={`${rIndex}-${cIndex}`} rIndex={rIndex} cIndex={cIndex} />)}
        </View>
    )
}

const s = StyleSheet.create({
    row: {
      flexDirection: 'row'
    }
})

export default Row;

