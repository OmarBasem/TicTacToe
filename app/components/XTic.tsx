import React from 'react';
import {Text, StyleSheet, Platform} from "react-native";

import {W} from '../utils'
import {theme} from "../theme";

const XTic = () => {
    return (
        <Text style={s.x}>X</Text>
    )
}

const s = StyleSheet.create({
    x: {
        color: theme.xColor,
        fontSize: W(25),
        bottom: Platform.OS === 'ios' ? 4 : 16
    }
})

export default XTic;