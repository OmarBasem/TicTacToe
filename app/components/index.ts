import Row from './Row';
import Box from './Box';
import XTic from './XTic';
import OTic from "./OTic";

export {
    Row,
    Box,
    XTic,
    OTic
}