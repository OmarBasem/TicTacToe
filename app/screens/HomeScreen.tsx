import React, {useReducer, useEffect} from 'react'
import {Text, View, StyleSheet, Alert} from 'react-native'

import {Row} from '../components'
import {W, findWinner} from "../utils";
import {theme} from "../theme";

export interface stateType {
    turn: string,
    ticks: Record<string, string>,
    xWins: number,
    oWins: number
}

const initialState = {turn: 'X', ticks: {}, xWins: 0, oWins: 0};

function reducer(state: stateType, action: Record<string, string>) {
    switch (action.type) {
        case 'X':
            return {...state, turn: 'O', ticks: {...state.ticks, [action.tickIndex]: action.type}};
        case 'O':
            return {...state, turn: 'X', ticks: {...state.ticks, [action.tickIndex]: action.type}};
        case 'PLAY_AGAIN':
            let xWins = state.xWins;
            let oWins = state.oWins;
            if (action.winner === 'X WIN')
                xWins += 1;
            else if (action.winner === 'O WIN')
                oWins += 1
            return {...state, ticks: {}, turn: 'X', xWins, oWins};
        default:
            throw new Error('Unknown action');
    }
}

export const TicksDispatch = React.createContext<any>(initialState);

const HomeScreen = () => {
    const rows = [0, 1, 2];
    const [state, dispatch] = useReducer(reducer, initialState);
    const xColor = state.turn === 'X' ? theme.xColor : theme.xColorDim
    const oColor = state.turn === 'O' ? theme.oColor : theme.oColorDim
    useEffect(() => {
        const winner = findWinner(state)
        if (winner !== 'CONTINUE') {
            Alert.alert('GAME OVER', winner, [{
                text: 'Play Again',
                onPress: () => dispatch({type: 'PLAY_AGAIN', tickIndex: '', winner})
            }])
        }
    }, [state.turn])
    return (
        <View style={s.view} testID='HomeScreen'>
            <View style={s.header}>
                <View>
                    <Text style={[s.tick, {color: xColor}]}>X</Text>
                    <View style={[s.line, {backgroundColor: state.turn === 'X' ? xColor : 'transparent'}]}/>
                    <Text style={[s.playerText, {color: xColor}]}>Player</Text>
                </View>
                <Text style={s.score}>{state.xWins} : {state.oWins}</Text>
                <View>
                    <Text style={[s.tick, {color: oColor}]}>O</Text>
                    <View style={[s.line, {backgroundColor: state.turn === 'O' ? oColor : 'transparent'}]}/>
                    <Text style={[s.playerText, {color: oColor}]}>Player</Text>
                </View>
            </View>
            <TicksDispatch.Provider value={{state, dispatch}}>
                {rows.map(rIndex => <Row key={rIndex.toString()} rIndex={rIndex}/>)}
            </TicksDispatch.Provider>
        </View>
    )

}


const s = StyleSheet.create({
    view: {
        backgroundColor: '#000000',
        flex: 1,
        alignItems: 'center',
        paddingTop: 60
    },
    header: {
        flexDirection: 'row',
        width: W(70),
        justifyContent: 'space-between',
        marginBottom: 40
    },
    line: {
        width: 60,
        height: 4,
        alignSelf: 'center'
    },
    playerText: {
        fontSize: 20,
        textAlign: 'center',
        marginTop: 8
    },
    tick: {
        fontSize: 40,
        textAlign: 'center'
    },
    score: {
        fontSize: 40,
        color: '#ffffff'
    }
})

export default HomeScreen;
