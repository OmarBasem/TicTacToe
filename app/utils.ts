import {Dimensions} from "react-native";
import {stateType} from "./screens/HomeScreen";

export const W = (percentage: number) => Dimensions.get('window').width / 100 * percentage;

export const findWinner = (state: stateType) => {
    const play = state.turn === 'X' ? 'O' : 'X';
    const {ticks} = state;
    const row0 = ticks['00'] === play && ticks['01'] === play && ticks['02'] === play
    const row1 = ticks['10'] === play && ticks['11'] === play && ticks['12'] === play
    const row2 = ticks['20'] === play && ticks['21'] === play && ticks['22'] === play
    const col0 = ticks['00'] === play && ticks['10'] === play && ticks['20'] === play
    const col1 = ticks['01'] === play && ticks['11'] === play && ticks['21'] === play
    const col2 = ticks['02'] === play && ticks['12'] === play && ticks['22'] === play
    const diagonalA = ticks['00'] === play && ticks['11'] === play && ticks['22'] === play
    const diagonalB = ticks['02'] === play && ticks['11'] === play && ticks['20'] === play
    const isWin = row0 || row1 || row2 || col0 || col1 || col2 || diagonalA || diagonalB
    if (isWin)
        return `${play} WIN`
    else if (Object.entries(ticks).length === 9)
        return 'DRAW'
    else
        return 'CONTINUE'
}