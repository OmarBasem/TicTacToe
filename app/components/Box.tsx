import React, {useContext} from 'react';
import {View, Text, StyleSheet, TouchableOpacity, Dimensions} from "react-native";

import XTic from "./XTic";
import OTic from "./OTic";
import {W} from '../utils'
import {TicksDispatch} from "../screens/HomeScreen";

const Box = ({rIndex, cIndex}: { rIndex: number, cIndex: number }) => {
    const {state, dispatch} = useContext(TicksDispatch);
    const tick = state.ticks[`${rIndex}${cIndex}`]
    return (
        <TouchableOpacity testID={`${rIndex}${cIndex}`} style={s.box} onPress={() => {
            if (!state.ticks[`${rIndex}${cIndex}`])
                dispatch({type: state.turn, tickIndex: `${rIndex}${cIndex}`})
        }}>
            {tick === 'X' ? <XTic/> : tick === 'O' ? <OTic/> : null}
        </TouchableOpacity>
    )
}


const s = StyleSheet.create({
    box: {
        width: W(30),
        height: W(30),
        borderWidth: 2,
        borderColor: '#ffffff',
        justifyContent: 'center',
        alignItems: 'center',
    },
})

export default Box;