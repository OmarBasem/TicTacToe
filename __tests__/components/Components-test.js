import React from 'react';
import renderer from "react-test-renderer";
import {XTic, OTic} from '../../app/components';
import {HomeScreen} from '../../app/screens'

describe('Testing components rendering', () => {

    test('HomeScreen renders correctly', () => {
        const tree = renderer.create(<HomeScreen />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    test('XTic renders correctly', () => {
        const tree = renderer.create(<XTic />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('OTic button renders correctly', () => {
        const tree = renderer.create(<OTic />).toJSON();
        expect(tree).toMatchSnapshot();
    });

})