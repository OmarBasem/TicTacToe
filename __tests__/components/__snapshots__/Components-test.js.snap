// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing components rendering HomeScreen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000000",
      "flex": 1,
      "paddingTop": 60,
    }
  }
  testID="HomeScreen"
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 40,
        "width": 525,
      }
    }
  >
    <View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 40,
              "textAlign": "center",
            },
            Object {
              "color": "rgb(239, 25, 27)",
            },
          ]
        }
      >
        X
      </Text>
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "height": 4,
              "width": 60,
            },
            Object {
              "backgroundColor": "rgb(239, 25, 27)",
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "fontSize": 20,
              "marginTop": 8,
              "textAlign": "center",
            },
            Object {
              "color": "rgb(239, 25, 27)",
            },
          ]
        }
      >
        Player
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontSize": 40,
        }
      }
    >
      0
       : 
      0
    </Text>
    <View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 40,
              "textAlign": "center",
            },
            Object {
              "color": "rgba(31, 237, 238, 0.5)",
            },
          ]
        }
      >
        O
      </Text>
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "height": 4,
              "width": 60,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "fontSize": 20,
              "marginTop": 8,
              "textAlign": "center",
            },
            Object {
              "color": "rgba(31, 237, 238, 0.5)",
            },
          ]
        }
      >
        Player
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="00"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="01"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="02"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="10"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="11"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="12"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="20"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="21"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#ffffff",
          "borderWidth": 2,
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID="22"
    />
  </View>
</View>
`;

exports[`Testing components rendering OTic button renders correctly 1`] = `
<Text
  style={
    Object {
      "bottom": 4,
      "color": "rgb(31, 237, 238)",
      "fontSize": 187.5,
    }
  }
>
  O
</Text>
`;

exports[`Testing components rendering XTic renders correctly 1`] = `
<Text
  style={
    Object {
      "bottom": 4,
      "color": "rgb(239, 25, 27)",
      "fontSize": 187.5,
    }
  }
>
  X
</Text>
`;
